### PR TITLE
WIP: Adapt Jenkinsfile to run mdbook tests

### LINKED ISSUE #1502: [MAINTENANCE] Update CI to test book
## Description

Concourse CI isn't testing the book ([concourse task](https://github.com/amethyst/amethyst/blob/master/.concourse/tasks/linux-test.yml)). This [used to be done](https://github.com/amethyst/amethyst/blob/master/.travis.yml#L70) in Travis CI.

## Reason

This keeps the book up to date as the code evolves.

## Impact

Would increase build times slightly.

### PR DESCRIPTION
**Pull request by marot**
## Description

This fixes #1502. I based the changes on https://github.com/amethyst/amethyst/commit/0e5b769553b3d7f67c37900a8450173760852887

I also fixed the book examples so that the CI would be successful.

## PR Checklist

By placing an x in the boxes I certify that I have:

- [ ] Ran `cargo test --all` locally if this modified any rs files.
- [ ] Ran `cargo +stable fmt --all` locally if this modified any rs files.
- [ ] Updated the content of the book if this PR would make the book outdated.
- [ ] Added a changelog entry if this will impact users, or modified more than 5 lines of Rust that wasn't a doc comment.
- [ ] Added unit tests for new APIs if any were added in this PR.
- [ ] Acknowledged that by making this pull request I release this code under an MIT/Apache 2.0 dual licensing scheme.

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -83,9 +83,17 @@ pipeline {
 			}
                     }
                     steps {
+		        echo 'Cleaning targets'
+			sh 'cargo clean'
                         echo 'Beginning tests...'
                         sh 'cargo test --all'
                         echo 'Tests done!'
+			echo 'Install mdbook'
```
Comment (azriel91) [inline Jenkinsfile]: Looks like the docker image [installs mdbook already](https://github.com/amethyst/builder/blob/9d8409287f885d0f8abe7807a7e31b4f2f5694e1/linux/Dockerfile#L32), so we don't have to.

Comment (marot) [reply Jenkinsfile]: Thanks for checking this

File: Jenkinsfile [hunk 2]
```diff
@@ -83,6 +83,8 @@ pipeline {
 			}
                     }
                     steps {
+			echo 'Build all the examples in the book'
+			sh 'mdbook test book -L target/debug/deps'
```
Comment (azriel91) [inline Jenkinsfile]: hm, i suspect this might have to go after `cargo test --all`, but i'll check with bors

**Comment by azriel91** [comment]
bors try

**Comment by marot** [comment]
Blocked by https://github.com/amethyst/amethyst/pull/1590

File: Jenkinsfile [hunk 3]
```diff
@@ -83,6 +83,8 @@ pipeline {
 			}
                     }
                     steps {
+			echo 'Build all the examples in the book'
```
Comment (LucioFranco) [inline Jenkinsfile]: we should make this its own stage probably

Comment (azriel91) [reply Jenkinsfile]: it needs the built libraries, but I don't wanna compile twice. Perhaps we make a shared artifact (or the Jenkins term for it) and copy the target directory contents across?

**Comment by azriel91** [comment]
two errors together are weird:

```
error[E0464]: multiple matching crates for `amethyst`
error[E0463]: can't find crate for `amethyst`
```

maybe we really do need a separate stage with `cargo clean` and `cargo build --all` before `mdbook test`

**Comment by marot** [comment]
@azriel91 Yep, I will try that this weekend. If you want to work on this earlier feel free to take this over :)

**Comment by torkleyy** [comment]
This is blocked on #1647

**Comment by torkleyy** [comment]
Ah, right, the old caching problem...

**Comment by torkleyy** [comment]
Closing since this is currently hard to do since `mdbook` can't deal with cached artifacts.

**Comment by azriel91** [comment]
wait, one last thing I wanna try, adding this before building:

```bash
rm -rf ./target/debug/deps/libamethyst* || true
```

**Comment by azriel91** [comment]
ah, I can't hijack the branch, gonna open another PR

**Pull request closed without merging**